### PR TITLE
Change the status of booking to Cancelled in case of no money on the account balance within Deadline

### PR DESCRIPTION
**Pull request by timur-mut**
https://track.happytravel.tech/issue/HP-114/Backend-task-for-issue-Change-the-status-of-booking-to-Cancelled-in-case-of-no-money-on-the-account-balance-within-Deadline

### CODE REVIEW

**Review by AlexeyShok — APPROVED** [review]

File: Api/Infrastructure/FunctionalExtensions/ResultTransactionExtensions.cs [hunk 1]
```diff
@@ -136,6 +161,32 @@ public static class ResultTransactionExtensions
             return await WithTransactionScope(context, () => f(result));
         }
 
+        public static async Task<Result<T, TE>> Bind<T, TE>(
+            this Task<Result<T>> target,
+            Func<T, Task<Result<T, TE>>> func,
+            Func<string, TE> f)
```
Comment (oleg-n-k) [inline Api/Infrastructure/FunctionalExtensions/ResultTransactionExtensions.cs:167]: fe ?

File: Api/Infrastructure/FunctionalExtensions/ResultTransactionExtensions.cs [hunk 2]
```diff
@@ -136,6 +161,32 @@ public static class ResultTransactionExtensions
...
+        {
+            var (_, isFailure, result, error) = await target;
+            return isFailure ? Result.Failure<T, TE>(f(error)) : await Result.Success<T, TE>(result).Bind(func);
+        }
+        
+        public static async Task<Result<T, TE>> Bind<T, TE>(
+            this Result target,
+            Func<Task<Result<T, TE>>> func,
+            Func<string, TE> f)
```
Comment (oleg-n-k) [inline Api/Infrastructure/FunctionalExtensions/ResultTransactionExtensions.cs:176]: fe

File: Api/Infrastructure/FunctionalExtensions/ResultTransactionExtensions.cs [hunk 3]
```diff
@@ -136,6 +161,32 @@ public static class ResultTransactionExtensions
...
+        {
+            var (_, isFailure, error) = target;
+            return isFailure ? Result.Failure<T, TE>(f(error)) : await func();
+        }
+
+        public static async Task<UnitResult<TE>> Bind<TE>(
+            this Result target,
+            Func<Task<UnitResult<TE>>> func,
+            Func<string, TE> f)
```
Comment (oleg-n-k) [inline Api/Infrastructure/FunctionalExtensions/ResultTransactionExtensions.cs:185]: fe

File: Api/Services/Accommodations/Bookings/BatchProcessing/BookingsProcessingService.cs [hunk 4]
```diff
@@ -118,17 +118,30 @@ async Task<Result<string>> Charge(Booking booking, ApiCaller apiCaller)
 
                 var chargeResult = await _accountPaymentService.Charge(booking, apiCaller);
 
-                if (chargeResult.IsFailure)
+                if (chargeResult.IsSuccess)
+                    return Result.Success(chargeResult.Value);
+
+                var reason = $"It is impossible charge money from a virtual account: {chargeResult.Error.Message}";
```
Comment (oleg-n-k) [inline Api/Services/Accommodations/Bookings/BatchProcessing/BookingsProcessingService.cs:124]: to charge

File: Api/Services/Accommodations/Bookings/BookingExecution/Flows/FinancialAccountBookingFlow.cs [hunk 5]
```diff
@@ -91,7 +91,8 @@ Task<Result<Data.Bookings.Booking>> RegisterBooking(BookingAvailabilityInfo book
             async Task<Result> ChargeMoney(Data.Bookings.Booking booking)
             {
                 var agentContext = await _agentContext.GetAgent();
-                return await _accountPaymentService.Charge(booking, agentContext.ToApiCaller());
+                var result = await _accountPaymentService.Charge(booking, agentContext.ToApiCaller());
+                return result.IsFailure ? Result.Failure(result.Error.Message) : Result.Success();
```
Comment (oleg-n-k) [inline Api/Services/Accommodations/Bookings/BookingExecution/Flows/FinancialAccountBookingFlow.cs:95]: return result.IsFailure 
? Result.Failure(result.Error.Message) 
: Result.Success();

File: HappyTravel.Edo.DirectApi/Services/Bookings/BookingCreationService.cs [hunk 6]
```diff
@@ -86,7 +86,8 @@ bool IsDeadlinePassed(Data.Bookings.Booking booking)
             async Task<Result> ChargeMoney(Data.Bookings.Booking booking)
             {
                 var agentContext = await _agentContext.GetAgent();
-                return await _accountPaymentService.Charge(booking, agentContext.ToApiCaller());
+                var result = await _accountPaymentService.Charge(booking, agentContext.ToApiCaller());
+                return result.IsFailure ? Result.Failure(result.Error.Message) : Result.Success();
```
Comment (oleg-n-k) [inline HappyTravel.Edo.DirectApi/Services/Bookings/BookingCreationService.cs:90]: return result.IsFailure 
? Result.Failure(result.Error.Message) 
: Result.Success();

**Review by oleg-n-k — APPROVED** [review]

**Pull request merged**